From 331b31055e155dca830716b0604c485705a9186b Mon Sep 17 00:00:00 2001
From: Tomas Henzl <thenzl@redhat.com>
Date: Mon, 8 Sep 2014 14:31:22 -0400
Subject: [scsi] hpsa: rescan devices on ioaccel2 error

Message-id: <1410186734-19986-32-git-send-email-thenzl@redhat.com>
Patchwork-id: 92041
O-Subject: [RHEL6.6 e-stor PATCH 31/83] hpsa: rescan devices on ioaccel2 error
Bugzilla: 1069185
RH-Acked-by: Maurizio Lombardi <mlombard@redhat.com>
RH-Acked-by: David Milburn <dmilburn@redhat.com>
RH-Acked-by: Ewan Milne <emilne@redhat.com>

Allow driver to schedule a rescan whenever a request fails on the ioaccel2 path.
This eliminates the possibility of driver getting stuck in non-ioaccel mode.

IOaccel mode (HP SSD Smart Path) is disabled by driver upon error detection.
Driver relied on idea that request would be retried through normal path, and a
subsequent error would occur on that path, and be processed by controller
firmware.  As part of that process, controller disables ioaccel mode and later
reinstates it, signalling driver to change modes.

In some error cases, the error will not duplicate on the standard path,
so the driver could get stuck in non-ioaccel mode.
To avoid that, we allow driver to request a rescan during the next run of the
rescan thread.

Signed-off-by: Scott Teel <scott.teel@hp.com>
Signed-off-by: Stephen M. Cameron <scameron@beardog.cce.hp.com>
Signed-off-by: James Bottomley <JBottomley@Parallels.com>
(cherry picked from commit e863d68e4881b080267ea2a6f97dfc658f0ad4fb)
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/scsi/hpsa.c b/drivers/scsi/hpsa.c
index 6ef2b2a..6b97761 100644
--- a/drivers/scsi/hpsa.c
+++ b/drivers/scsi/hpsa.c
@@ -1475,6 +1475,7 @@ static void process_ioaccel2_completion(struct ctlr_info *h,
     "%s: Error 0x%02x, Retrying on standard path.\n",
     "HP SSD Smart Path", c2->error_data.status);
   dev->offload_enabled = 0;
+  h->drv_req_rescan = 1; /* schedule controller for a rescan */
   cmd->result = DID_SOFT_ERROR << 16;
   cmd_free(h, c);
   cmd->scsi_done(cmd);
@@ -6176,6 +6177,9 @@ static int hpsa_kickoff_rescan(struct ctlr_info *h)
  int i;
  char *event_type;
 
+ /* Clear the driver-requested rescan flag */
+ h->drv_req_rescan = 0;
+
  /* Ask the controller to clear the events we're handling. */
  if ((h->transMethod & (CFGTBL_Trans_io_accel1
    | CFGTBL_Trans_io_accel2)) &&
@@ -6223,7 +6227,9 @@ static int hpsa_kickoff_rescan(struct ctlr_info *h)
 
 /* Check a register on the controller to see if there are configuration
  * changes (added/changed/removed logical drives, etc.) which mean that
- * we should rescan the controller for devices.  If so, add the controller
+ * we should rescan the controller for devices.
+ * Also check flag for driver-initiated rescan.
+ * If either flag or controller event indicate rescan, add the controller
  * to the list of controllers needing to be rescanned, and gets a
  * reference to the associated scsi_host.
  */
@@ -6233,7 +6239,7 @@ static void hpsa_ctlr_needs_rescan(struct ctlr_info *h)
   return;
 
  h->events = readl(&(h->cfgtable->event_notify));
- if (!h->events)
+ if (!h->events && !h->drv_req_rescan)
   return;
 
  /*
@@ -6407,6 +6413,8 @@ reinit_after_soft_reset:
  /* Enable Accelerated IO path at driver layer */
  h->acciopath_status = 1;
 
+ h->drv_req_rescan = 0;
+
  /* Turn the interrupts on so we can service requests */
  h->access.set_intr_mask(h, HPSA_INTR_ON);
 
diff --git a/drivers/scsi/hpsa.h b/drivers/scsi/hpsa.h
index 1d3340d..2e7521f 100644
--- a/drivers/scsi/hpsa.h
+++ b/drivers/scsi/hpsa.h
@@ -182,6 +182,7 @@ struct ctlr_info {
 #define HPSATMF_LOG_QRY_ASYNC   (1 << 25)
  u32 events;
  int acciopath_status;
+ int drv_req_rescan; /* flag for driver to request rescan event */
 };
 #define HPSA_ABORT_MSG 0
 #define HPSA_DEVICE_RESET_MSG 1
-- 
1.7.1